### PR TITLE
BR: flush causal timestamp before backup start for rawkv apiv2

### LINKED ISSUE #12989: Flush causal timestamp before backup start for RawKV apiv2
## Development Task

Flush causal timestamp before backup start for RawKV apiv2

see issue: https://github.com/tikv/migration/issues/138

### PR DESCRIPTION
**Pull request by haojinming**
### What is changed and how it works?

Issue Number: Close #12989 

What's Changed:

1. Flush causal timestamp before backup start for rawkv apiv2, see issue: https://github.com/tikv/migration/issues/138
2. Address review comments: https://github.com/tikv/tikv/pull/12866#discussion_r904483116
   let cdc::Endpoint hold the causal ts provider instead of cdc::Observer. 

### Check List 

Tests 

- Unit test
- Manual test (add detailed scripts or steps below)

Side effects

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/backup/src/endpoint.rs [hunk 1]
```diff
@@ -962,6 +966,24 @@ impl<E: Engine, R: RegionInfoProvider + Clone + 'static> Endpoint<E, R> {
             }
             return;
         }
+        // CausalTsProvider may cache tso. Flush the cached tso to avoid use them in future data.
```
Comment (pingyu) [inline components/backup/src/endpoint.rs:969]: Suggest: Flush causal timestamp to make sure that future writes will have larger timestamps, and help TiKV-BR acuqire a `backup-ts` with intact data smaller than it. (Note that intactness is not fully ensured now, until the `safe-ts` of RawKV is implemented. TiKV-BR need a workaround by rewinding `backup-ts` to a small "safe interval").

File: components/cdc/src/endpoint.rs [hunk 2]
```diff
@@ -2226,6 +2240,22 @@ mod tests {
...
+        let sleep_interval = Duration::from_secs(1);
+        let cfg = CdcConfig {
+            min_ts_interval: ReadableDuration(sleep_interval),
+            ..Default::default()
+        };
+        let ts_provider = Arc::new(causal_ts::tests::TestProvider::default());
+        let start_ts = ts_provider.get_ts().unwrap();
+        let _suite =
+            mock_endpoint_with_ts_provider(&cfg, None, ApiVersion::V2, Some(ts_provider.clone()));
+        std::thread::sleep(sleep_interval * 2);
```
Comment (pingyu) [inline components/cdc/src/endpoint.rs:2254]: How about test by schedule `Task::MinTS` as the above test case ? The sleep & wait will not be stable in a busy test infra.

File: components/backup/src/endpoint.rs [hunk 3]
```diff
@@ -962,6 +966,24 @@ impl<E: Engine, R: RegionInfoProvider + Clone + 'static> Endpoint<E, R> {
...
+        // BR assume that all data with smaller ts is backup-ed.
+        if request.is_raw_kv {
```
[the remark this answers is not in the document]
Comment (haojinming) [reply components/backup/src/endpoint.rs:971]: `causal_ts_provider` is set to none when apiversion == v1/v1ttl. Should we check apiversion again?

File: components/server/src/server.rs [hunk 4]
```diff
@@ -334,6 +336,7 @@ impl<ER: RaftEngine> TiKvServer<ER> {
             flow_info_receiver: None,
             sst_worker: None,
             quota_limiter,
+            causal_ts_provider: None,
```
Comment (zeminzhou) [inline components/server/src/server.rs:339]: Why add causal_ts_provider into TiKvServer?

**Comment by zeminzhou** [comment]
LGTM~

File: components/server/src/server.rs (hunk 4 above)
Comment (haojinming) [reply components/server/src/server.rs:339]: We need pass `causal_ts_provider` to `cdc::Endpoint` and `backup::Endpoint`. `cdc::Endpoint` is created later in this function, but cannot pass an Optional variable here as lifecyle compile issue and `backup::Endpoint` is created in another function. So store `causal_ts_provider` in TiKVServer is the easiest and reasonable way I found.

File: components/server/src/server.rs [hunk 5]
```diff
@@ -815,9 +818,10 @@ impl<ER: RaftEngine> TiKvServer<ER> {
             }
             let causal_ts_provider = Arc::new(tso.unwrap());
             info!("Causal timestamp provider startup.");
-            cdc_ob.set_causal_ts_provider(causal_ts_provider.clone());
-            let causal_ob = causal_ts::CausalObserver::new(causal_ts_provider, cdc_ob.clone());
+            let causal_ob =
+                causal_ts::CausalObserver::new(causal_ts_provider.clone(), cdc_ob.clone());
             causal_ob.register_to(self.coprocessor_host.as_mut().unwrap());
+            self.causal_ts_provider = Some(causal_ts_provider);
```
[the remark this answers is not in the document]
Comment (haojinming) [reply components/server/src/server.rs]: Moved `causal_ts_provider` creation to `TiKVServer::init` as it's more make sense to init member variable in init process.

**Review by pingyu — APPROVED** [review]
LGTM~

**Review by Connor1996 — APPROVED** [review]
LGTM

**Review by overvenus — APPROVED** [review]

**Pull request merged**